File: Sources/TBKit/Chain.swift
import Foundation

public enum PendingStatus: String, Sendable, Hashable {
    case posted, voided, expired, pending
    /// Nothing resolved it within the scanned window and the scan did not reach the account's newest transfer.
    case unknown
}

public struct PendingResolution: Sendable, Hashable {
    public var status: PendingStatus
    /// Post/void transfers that reference the pending transfer.
    public var resolutions: [Transfer]
    /// Later debit-account transfers examined.
    public var scanned: UInt32
    /// True when the scan reached the newest transfer on the debit account.
    public var exhausted: Bool
    /// Expiry in ns since epoch; nil when the pending never expires.
    public var expiresAt: UInt64?
}

public struct Chain: Sendable, Hashable {
    public var transfer: Transfer
    /// The referenced pending transfer, when `transfer` posts or voids one.
    public var pending: Transfer?
    /// Set when `transfer` or the pending it references is a pending transfer.
    public var resolution: PendingResolution?
    /// Members of the linked group containing `transfer`, in timestamp order; empty if not linked.
    public var linked: [Transfer]
}

extension TBClient {
    public static let defaultLookback: UInt32 = 1_000
    public static let maxLookback: UInt32 = 1_000_000

    /// Resolves pending → post/void relationships and the linked group for a transfer.
    public func chain(for id: UInt128, lookback: UInt32 = defaultLookback, now: Date = .now) async throws(TBError) -> Chain {
        let t = try await lookupTransfer(id)
        var chain = Chain(transfer: t, pending: nil, resolution: nil, linked: [])
        let lookback = min(max(lookback, 1), Self.maxLookback)

        var pending: Transfer? = t.flags.contains(.pending) ? t : nil
        if t.pendingID != 0, let p = try await lookupTransfers([t.pendingID]).first {
            chain.pending = p
            pending = p
        }
        if let pending {
            chain.resolution = try await resolvePending(pending, lookback: lookback, now: now)
        }

        let group = try await linkedGroup(t)
        if group.count > 1 { chain.linked = group }
        return chain
    }

    /// Post/void transfers debit the same account as the pending one and follow it in time,
    /// so scan that account's later debits for a matching `pending_id`.
    func resolvePending(_ p: Transfer, lookback: UInt32, now: Date) async throws(TBError) -> PendingResolution {
        var res = PendingResolution(status: .unknown, resolutions: [], scanned: 0, exhausted: false, expiresAt: nil)
        if p.timeout > 0 {
            res.expiresAt = p.timestamp &+ UInt64(p.timeout) &* 1_000_000_000
        }

        var cursor = p.timestamp &+ 1
        while res.scanned < lookback {
            let page = min(lookback - res.scanned, tbMaxLimit)
            let batch = try await accountTransfers(
                p.debitAccountID,
                AccountFilter(debits: true, credits: false, timestampMin: cursor, limit: page))
            res.scanned += UInt32(batch.count)
            res.resolutions.append(contentsOf: batch.filter { $0.pendingID == p.id })
            if !res.resolutions.isEmpty { break }
            if batch.count < page {
                res.exhausted = true
                break
            }
            cursor = batch[batch.count - 1].timestamp &+ 1
        }

        if res.resolutions.contains(where: { $0.flags.contains(.postPendingTransfer) }) {
            res.status = .posted
        } else if res.resolutions.contains(where: { $0.flags.contains(.voidPendingTransfer) }) {
            res.status = .voided
        } else if let exp = res.expiresAt, TBFormat.nanos(now) >= exp {
            res.status = .expired
        } else if res.exhausted {
            res.status = .pending
        }
        return res
    }

    /// Events in one batch receive consecutive timestamps and a linked chain ends at the first
    /// event without the `linked` flag. An unfiltered `query_transfers` walks the global
    /// timestamp index, so contiguity is checked against every transfer.
    func linkedGroup(_ t: Transfer) async throws(TBError) -> [Transfer] {
        var group = [t]
        let pageLimit: UInt32 = 8189 // a linked chain never exceeds one batch

        if t.timestamp > 1 {
            let before = try await queryTransfers(
                QueryFilter(timestampMin: 1, timestampMax: t.timestamp - 1, limit: pageLimit, reversed: true))
            var expect = t.timestamp - 1
            for b in before {
                guard b.timestamp == expect, b.flags.contains(.linked) else { break }
                group.append(b)
                expect -= 1
            }
        }

        if t.flags.contains(.linked) {
            let after = try await queryTransfers(QueryFilter(timestampMin: t.timestamp + 1, limit: pageLimit))
            var expect = t.timestamp + 1
            for a in after {
                guard a.timestamp == expect else { break }
                group.append(a)
                if !a.flags.contains(.linked) { break }
                expect += 1
            }
        }
        return group.sorted { $0.timestamp < $1.timestamp }
    }
}

File: cli/tbclient/src/chain.rs
//! Pending → post/void resolution, and linked groups.
//!
//! TigerBeetle stores neither relationship as a link you can follow: a post or void names the
//! pending transfer it resolves, and a linked group is contiguous timestamps with a flag. Both are
//! therefore reconstructed by scanning, exactly as `ui/Sources/TBKit/Chain.swift` does.

use std::time::{SystemTime, UNIX_EPOCH};

use crate::client::{Client, TbError};
use crate::models::{AccountFilter, MAX_LIMIT, QueryFilter, Transfer, transfer_flags};

pub const DEFAULT_LOOKBACK: u32 = 1_000;
pub const MAX_LOOKBACK: u32 = 1_000_000;

/// A linked chain never exceeds one batch.
const LINKED_PAGE: u32 = 8_189;

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum PendingStatus {
    Posted,
    Voided,
    Expired,
    Pending,
    /// Nothing resolved it within the scanned window, and the scan did not reach the account's
    /// newest transfer — so the answer is "not known from here", not "still pending".
    Unknown,
}

impl PendingStatus {
    pub fn label(&self) -> &'static str {
        match self {
            Self::Posted => "posted",
            Self::Voided => "voided",
            Self::Expired => "expired",
            Self::Pending => "pending",
            Self::Unknown => "unresolved in lookback",
        }
    }
}

#[derive(Clone, Debug)]
pub struct PendingResolution {
    pub status: PendingStatus,
    /// Post or void transfers that reference the pending transfer.
    pub resolutions: Vec<Transfer>,
    /// Later debit-account transfers examined.
    pub scanned: u32,
    /// True when the scan reached the newest transfer on the debit account.
    pub exhausted: bool,
    /// Expiry in nanoseconds since the epoch; `None` when the pending never expires.
    pub expires_at: Option<u64>,
}

#[derive(Clone, Debug)]
pub struct Chain {
    pub transfer: Transfer,
    /// The referenced pending transfer, when this one posts or voids.
    pub pending: Option<Transfer>,
    /// Set when this transfer, or the pending it references, is a pending transfer.
    pub resolution: Option<PendingResolution>,
    /// Members of the linked group in timestamp order; empty when the transfer is not linked.
    pub linked: Vec<Transfer>,
}

fn now_nanos() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_nanos() as u64)
        .unwrap_or(0)
}

impl Client {
    pub fn chain(&self, id: u128, lookback: u32) -> Result<Chain, TbError> {
        self.chain_at(id, lookback, now_nanos())
    }

    /// `now` is a parameter so expiry is testable without waiting for a clock.
    pub fn chain_at(&self, id: u128, lookback: u32, now: u64) -> Result<Chain, TbError> {
        let transfer = *self
            .lookup_transfers(&[id])?
            .first()
            .ok_or_else(|| TbError::NotFound(format!("transfer {id}")))?;

        let lookback = lookback.clamp(1, MAX_LOOKBACK);
        let mut chain = Chain {
            transfer,
            pending: None,
            resolution: None,
            linked: Vec::new(),
        };

        let mut pending = (transfer.flags & transfer_flags::PENDING != 0).then_some(transfer);
        if transfer.pending_id != 0
            && let Some(referenced) = self.lookup_transfers(&[transfer.pending_id])?.first()
        {
            chain.pending = Some(*referenced);
            pending = Some(*referenced);
        }
        if let Some(pending) = pending {
            chain.resolution = Some(self.resolve_pending(&pending, lookback, now)?);
        }

        let group = self.linked_group(&transfer)?;
        if group.len() > 1 {
            chain.linked = group;
        }
        Ok(chain)
    }

    /// A post or void debits the same account as its pending transfer and follows it in time, so
    /// the search is a bounded forward scan of that account's debits.
    fn resolve_pending(
        &self,
        pending: &Transfer,
        lookback: u32,
        now: u64,
    ) -> Result<PendingResolution, TbError> {
        let mut out = PendingResolution {
            status: PendingStatus::Unknown,
            resolutions: Vec::new(),
            scanned: 0,
            exhausted: false,
            expires_at: (pending.timeout > 0).then(|| {
                pending
                    .timestamp
                    .wrapping_add(pending.timeout as u64 * 1_000_000_000)
            }),
        };

        let mut cursor = pending.timestamp.wrapping_add(1);
        while out.scanned < lookback {
            let page = (lookback - out.scanned).min(MAX_LIMIT);
            let batch = self.account_transfers(&AccountFilter {
                account_id: pending.debit_account_id,
                debits: true,
                credits: false,
                timestamp_min: cursor,
                limit: page,
                ..Default::default()
            })?;
            out.scanned += batch.len() as u32;
            out.resolutions
                .extend(batch.iter().filter(|t| t.pending_id == pending.id).copied());
            if !out.resolutions.is_empty() {
                break;
            }
            if (batch.len() as u32) < page {
                out.exhausted = true;
                break;
            }
            cursor = batch[batch.len() - 1].timestamp.wrapping_add(1);
        }

        out.status = if out
            .resolutions
            .iter()
            .any(|t| t.flags & transfer_flags::POST_PENDING_TRANSFER != 0)
        {
            PendingStatus::Posted
        } else if out
            .resolutions
            .iter()
            .any(|t| t.flags & transfer_flags::VOID_PENDING_TRANSFER != 0)
        {
            PendingStatus::Voided
        } else if out.expires_at.is_some_and(|expiry| now >= expiry) {
            PendingStatus::Expired
        } else if out.exhausted {
            PendingStatus::Pending
        } else {
            PendingStatus::Unknown
        };
        Ok(out)
    }

    /// Events in one batch get consecutive timestamps, and a linked chain ends at the first event
    /// without the flag. An unfiltered query walks the global timestamp index, so contiguity has
    /// to be checked against every transfer, not just this account's.
    fn linked_group(&self, transfer: &Transfer) -> Result<Vec<Transfer>, TbError> {
        let mut group = vec![*transfer];

        if transfer.timestamp > 1 {
            let before = self.query_transfers(&QueryFilter {
                timestamp_min: 1,
                timestamp_max: transfer.timestamp - 1,
                limit: LINKED_PAGE,
                reversed: true,
                ..Default::default()
            })?;
            let mut expect = transfer.timestamp - 1;
            for candidate in before {
                if candidate.timestamp != expect || candidate.flags & transfer_flags::LINKED == 0 {
                    break;
                }
                group.push(candidate);
                expect -= 1;
            }
        }

        if transfer.flags & transfer_flags::LINKED != 0 {
            let after = self.query_transfers(&QueryFilter {
                timestamp_min: transfer.timestamp + 1,
                limit: LINKED_PAGE,
                ..Default::default()
            })?;
            let mut expect = transfer.timestamp + 1;
            for candidate in after {
                if candidate.timestamp != expect {
                    break;
                }
                group.push(candidate);
                if candidate.flags & transfer_flags::LINKED == 0 {
                    break;
                }
                expect += 1;
            }
        }

        group.sort_by_key(|t| t.timestamp);
        Ok(group)
    }
}
